### PR TITLE
Draft for further discussion.

### PR DESCRIPTION
**Pull request by amitjamadagni**
Please do review and comment. Thanks.

### CODE REVIEW

File: GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+* GitHub Profile : https://github.com/amitjamadagni
+* GSoC 2014 : https://www.google-melange.com/gsoc/project/details/google/gsoc2014/amitjamadagni/5676830073815040
+* GSoC 2014 blog : https://knotsknotted.wordpress.com/category/gsoc/
+* Contribution to Julia : Implementation of ror! and rol! (https://github.com/JuliaLang/julia/pull/9822)         
+* Contribution to Julia Quantum : Normalize to QuArray (https://github.com/JuliaQuantum/QuBase.jl/pull/15)           
+
+---
+### Project Synopsis : The main aim of the project is to achieve 
+
+1. Framework to solve SE or ME by
```
Comment (garrison) [inline GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md]: What do these stand for?  My best guess is "stochastic equation" and "master equation"?

Comment (garrison) [inline GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md]: or perhaps SE is really Schrodinger equation.  Would be best to expand these abbreviations, I think.

**Comment by amitjamadagni** [comment]
@garrison Thanks for the comments. This was an initial draft and I am currently working on the changes and should be sending in a commit soon with more detailing.  @acroy has been helping me with this and I hope that by the next commit it would be more neat and clear :).

**Comment by amitjamadagni** [comment]
@acroy I have worked on the edits. A review would be helpful. Thanks !!!!

File: GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+---
+### Project Synopsis : The main inspiration is to equip the library with solvers which are on par with QuTiP. Main aims of the project include : 
+
+1. Framework to solve Schrodinger (SE) and Quantum Master Equations (ME) by
+    * Exponential solvers
+    * ODE solvers (RK Methods)
+    * Monte Carlo Wave Function Method (MCWF)
+    * Steady State Methods
+
+2. Paralleling some of the above tasks (Integrating QuArrays and methods involved in MCWF) using the existing infrastructure for parallel computing features in Julia to achieve greater efficiency.
```
Comment (acroy) [inline GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md]: Parallelizing ...

File: GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+3. In the course of integrating other packages, contribute back to improve their functionality and usability.
+
+---
+# Details :
```
Comment (acroy) [inline GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md]: `##Details:` and the following section titles should be at most `##`.

**Comment by acroy** [comment]
@garrison Would be great if you could have a second, independent look at the proposal. Then we can see which parts have to be written more clearly.

**Comment by garrison** [comment]
I would suggest expanding `SE` and `ME` in the type names as well.  I would also expand RK in the proposal, even though most people will know what it means.

Is it possible to come up with a more specific name for `QuantumMonteCarlo`?  "Quantum Monte Carlo" can mean a variety of different things.

I think the project looks ambitious, and look forward to seeing the results.

I am not incredibly familiar with QuTiP (or open quantum systems, for that matter), so I am not the best judge of the contents of the proposal.

One question though: from what I can tell QuTiP seems to have no mechanism for dealing with fermions (i.e. it seems to assume that everything is bosonic).  Do you have any thoughts on how fermions ought to be implemented?  The reason I ask is because I'm wondering if it's something that should be considered at a very early (design) stage when the groundwork is being constructed.

**Comment by acroy** [comment]
@garrison : Thanks for your comments! It's good to get a fresh pair of eyes on this.

> Is it possible to come up with a more specific name for QuantumMonteCarlo? "Quantum Monte Carlo" can mean a variety of different things.

Yes, it is better to call it MCWF (Monte Carlo Wave Function method), which is more accurate.

> Do you have any thoughts on how fermions ought to be implemented?

You are right regarding QuTiP. To be honest I haven't thought about the fermions yet. One possibility would be to use products of two/four level states and having some way to restrict those to a certain sub-space ... but this is probably not the most effective way. I know that @Jutho is working on things related to MPS and we have discussed if/how we can use his work together with `QuArrays` in JuliaQuantum/QuBase.jl#2.

**Comment by amitjamadagni** [comment]
@garrison @acroy Thanks for the insight and comments. I will try making the appropriate changes. I still need to add some more details to the current version, the pace is a bit slow as I have been having exams at the university. I should be adding more detail soon. I just have gone through the comments on the issue, and it seems the `AbstractQuArray` and `AbstractArray` are facilitating the necessary implementation,  please do correct me if I am getting something wrong as my knowledge is a bit limited here.  
It would be really helpful to know if anything more specific could be done with respect to the current state of the proposal. Hoping to hear from you on this and also the next steps to be taken. Thanks.

File: GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+propagate(qe::QuantumEquation, init_state::QuState, tlist, method=:RK)
+       # this would return an instance of RK method for quantum state evolution.
+``` 
+One of the aims would be to use Julia's iterator interface to get states at intermediate steps.
+
+### Prototypes :
+
+At the end of the project, the results would look as follows :
+
+Birth and Death of Photons in a Cavity.
```
Comment (acroy) [inline GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md]: Maybe put this as a comment on top of the code below?

File: GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+ 
+Ref : QuTiP Example [here] (http://nbviewer.ipython.org/github/qutip/qutip-notebooks/blob/master/examples/example-photon_birth_death.ipynb) 
+
+```ijulia
+
+a = lowerop(5)
```
Comment (acroy) [inline GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md]: We should put some more comments here:

```
a = lowerop(5) # annihilates a photon in cavity mode
H = a'*a   # Hamiltonian of cavity mode
init_state = densitymatrix(5) # create initial density matrix (ground state)
kappa = 1.0/0.129 # coupling strength to thermal bath
nth = 0.063      # thermal occupation of thermal bath

# construct Lindblad operators
c_ops = [sqrt(kappa * (1 + nth)) * a, sqrt(kappa * nth) * a']
```

File: GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+ 
+* GitHub Profile : https://github.com/amitjamadagni
+* GSoC 2014 : https://www.google-melange.com/gsoc/project/details/google/gsoc2014/amitjamadagni/5676830073815040
+* GSoC 2014 blog : https://knotsknotted.wordpress.com/category/gsoc/
+* Contribution to Julia : Implementation of ror! and rol! (https://github.com/JuliaLang/julia/pull/9822)         
+* Contribution to Julia Quantum : Normalize to QuArray (https://github.com/JuliaQuantum/QuBase.jl/pull/15)           
+
+---
+### Project Synopsis : The main inspiration is to equip the library with solvers which are on par with QuTiP.
```
Comment (acroy) [inline GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md]: I would put the "main inspiration as a first sentence in this section and extend it slightly. How about

The aim of the project is to equip the JuliaQuantum libraries with solvers which are on par with QuTiP. To this end a flexible and unified framework will be created, which allows to access different solvers. The project will make use of existing packages (if possible) and aims to contribute back in order to homogenize their interfaces and improve the performance.

Main goals of the project include :

File: GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+* Monte Carle Wavefunction : The implementation would be based on the proposed implementation in QuDOS.jl package and also with reference from QuTiP.
+
+* Steady State equation : Various methods based out of QuTiP. 
+
+The parallelization techniques can be achieved at certain level of abstraction. Some of the ideas are as follows :
+
+*  Level of Abstraction --> Algorithmic (for example : Monte Carlo methods)
+
+*  Level of Abstraction --> Type definitions (for example : `QuArray` based out of DArray or using the package [ParallelSparseMatMul.jl] (https://github.com/madeleineudell/ParallelSparseMatMul.jl) given the fact `QuArray`s are based on arrays.
+
```
Comment (acroy) [inline GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md:64]: Should we have 1-2 sentences here to summarize what we will have at the end of the project? Something like

At the end of the project, JuliaQuantum will be equipped with a framework to solve dynamical equations arising in quantum mechanics. The framework itself and the structures defined in QuBase will be useable with distributed/shared arrays to enable development of methods using parallel computing.

File: GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md (hunk 6 above)
Comment (acroy) [inline GSoC/2015/Julia-Quantum-:-Framework-for-Solvers.md]: Maybe JuliaQuantum should be a link to our webpage?

**Comment by acroy** [comment]
@amitjamadagni I just added some more comments, but otherwise I think the proposal looks Ok. I would make the changes, sleep over it and if there are no objections submit it according to the numfocus guidelines.

**Comment by acroy** [comment]
>  I just have gone through the comments on the issue, and it seems the AbstractQuArray and AbstractArray are facilitating the necessary implementation

The problem was that initially the coefficient container was `<:AbstractArray`, but @Jutho's tensors are not (if I remember correctly). But you are right, the current implementation should basically allow to use "non-AbstractArrays" as containers ...

**Comment by amitjamadagni** [comment]
@acroy I have created a pull request on the numfocus repo. Thanks once again @acroy @garrison  for all the help in making the proposal look better :)

**Comment by i2000s** [comment]
Good luck on this!

**Comment by acroy** [comment]
@amitjamadagni : Good luck and don't forget to submit the proposal to GSoC :-)

**Comment by garrison** [comment]
Regarding fermions, let me a elaborate a bit on my concern.  Right now you create multiple bosonic (or spin) modes in QuTiP by taking a [tensor product](http://qutip.org/docs/3.1.0/guide/guide-tensor.html) of the different modes.  This works well for bosons and spins, where different modes commute, but it does not work for fermions, as a creation operator on one site has a phase ($\pm 1$) which depends on which other modes are occupied, due to the fact that the creation operators on two different sites must anti-commute.

I _think_ it would be possible to make QuTiP work with fermions by simply making the tensor product and partial trace functions work differently for fermions (i.e. have them deal with these +/- signs automatically, albeit at some loss of efficiency).

In any event, even if fermions are possible with minor changes (and I am not 100% sure they are), it may be good (before commencing the actual project) to reach out to the QuTiP developers to ask if there are any design elements they would do differently, if they were starting again from scratch.

**Comment by acroy** [comment]
@garrison You are of course right regarding the anticommutativity. It is probably not impossible to keep track of the signs, but doing this efficiently might be hard. In any case we should do some research on this problem and maybe contact the QuTiP authors. Could you open an issue for QuBase.jl so we don't forget about this?

**Comment by amitjamadagni** [comment]
@i2000s thanks for all the help. @acroy I have submitted the proposal on melange. Thanks once again :).

**Comment by amitjamadagni** [comment]
@i2000s @garrison I am very happy to inform that the project has been selected as a part of GSoC 2015 and I would be working under the guidance of @acroy . I would like to thank you all for the support and guidance, hoping to have great summer of code ahead. :)

**Comment by garrison** [comment]
Happy to hear!

**Comment by i2000s** [comment]
@amitjamadagni I have heard of this excellent piece of news from @acroy!  Congratulations on what you have achieved through your passionate efforts! 

Do you want me to merge this proposal to the Resources repo for future reference, or just close it? Now, it is time to inform the community your progress and thoughts on the JuliaQuantum website. I will email you more details...

**Comment by jrevels** [comment]
This is definitely awesome, congrats! Looking forward to this summer.

**Comment by amitjamadagni** [comment]
@i2000s @jrevels :)

@i2000s it would be great if we could merge this as it would help for future reference. Thanks for the email on how to progress with the updates. I will send in a pull with the update.

**Comment by amitjamadagni** [comment]
@i2000s @jrevels @garrison I am happy to inform that we are successfully done with GSoC 2015 and the results have been reflected in the package [QuDynamics.jl](https://github.com/JuliaQuantum/QuDynamics.jl). Thank you all once again for all the help and guidance !

**Comment by i2000s** [comment]
Congrats!

**Comment by jrevels** [comment]
:100:

**Pull request merged**